<a id="load"></a>
## 1. Load and explore claims data

The csv file `claimsdata.csv` Load the file to Apache® Spark DataFrame using code below.

In [ ]:
import ibmos2spark

In [4]:
# The code was removed by Watson Studio for sharing.

[Row(DESYNPUF_ID='00016F745862898F', CLM_ID='5.42272E+14', SEGMENT='1', CLM_FROM_DT='20090602', CLM_THRU_DT='20090602', PRVDR_NUM='3901GS', CLM_PMT_AMT='30', NCH_PRMRY_PYR_CLM_PD_AMT='0', AT_PHYSN_NPI='2963419753', OP_PHYSN_NPI=None, OT_PHYSN_NPI='2963419753', NCH_BENE_BLOOD_DDCTBL_LBLTY_AM='0', ICD9_DGNS_CD_1='V5832', ICD9_DGNS_CD_2='V5861', ICD9_DGNS_CD_3='2724', ICD9_DGNS_CD_4='3182', ICD9_DGNS_CD_5='V5869', ICD9_DGNS_CD_6='42731', ICD9_DGNS_CD_7=None, ICD9_DGNS_CD_8=None, ICD9_DGNS_CD_9=None, ICD9_DGNS_CD_10=None, ICD9_PRCDR_CD_1=None, ICD9_PRCDR_CD_2=None, ICD9_PRCDR_CD_3=None, ICD9_PRCDR_CD_4=None, ICD9_PRCDR_CD_5=None, ICD9_PRCDR_CD_6=None, NCH_BENE_PTB_DDCTBL_AMT='0', NCH_BENE_PTB_COINSRNC_AMT='0', ADMTNG_ICD9_DGNS_CD=None, HCPCS_CD_1='85610', HCPCS_CD_2='80048', HCPCS_CD_3='80061', HCPCS_CD_4='82306', HCPCS_CD_5='96372', HCPCS_CD_6='87088', HCPCS_CD_7='85025', HCPCS_CD_8='80076', HCPCS_CD_9='84075', HCPCS_CD_10=None, HCPCS_CD_11=None, HCPCS_CD_12=None, HCPCS_CD_13=None, HCPCS_

In [5]:
df_data.printSchema()

root
 |-- DESYNPUF_ID: string (nullable = true)
 |-- CLM_ID: string (nullable = true)
 |-- SEGMENT: string (nullable = true)
 |-- CLM_FROM_DT: string (nullable = true)
 |-- CLM_THRU_DT: string (nullable = true)
 |-- PRVDR_NUM: string (nullable = true)
 |-- CLM_PMT_AMT: string (nullable = true)
 |-- NCH_PRMRY_PYR_CLM_PD_AMT: string (nullable = true)
 |-- AT_PHYSN_NPI: string (nullable = true)
 |-- OP_PHYSN_NPI: string (nullable = true)
 |-- OT_PHYSN_NPI: string (nullable = true)
 |-- NCH_BENE_BLOOD_DDCTBL_LBLTY_AM: string (nullable = true)
 |-- ICD9_DGNS_CD_1: string (nullable = true)
 |-- ICD9_DGNS_CD_2: string (nullable = true)
 |-- ICD9_DGNS_CD_3: string (nullable = true)
 |-- ICD9_DGNS_CD_4: string (nullable = true)
 |-- ICD9_DGNS_CD_5: string (nullable = true)
 |-- ICD9_DGNS_CD_6: string (nullable = true)
 |-- ICD9_DGNS_CD_7: string (nullable = true)
 |-- ICD9_DGNS_CD_8: string (nullable = true)
 |-- ICD9_DGNS_CD_9: string (nullable = true)
 |-- ICD9_DGNS_CD_10: string (nullable = 

In [6]:
df_data.take(1)

[Row(DESYNPUF_ID='00016F745862898F', CLM_ID='5.42272E+14', SEGMENT='1', CLM_FROM_DT='20090602', CLM_THRU_DT='20090602', PRVDR_NUM='3901GS', CLM_PMT_AMT='30', NCH_PRMRY_PYR_CLM_PD_AMT='0', AT_PHYSN_NPI='2963419753', OP_PHYSN_NPI=None, OT_PHYSN_NPI='2963419753', NCH_BENE_BLOOD_DDCTBL_LBLTY_AM='0', ICD9_DGNS_CD_1='V5832', ICD9_DGNS_CD_2='V5861', ICD9_DGNS_CD_3='2724', ICD9_DGNS_CD_4='3182', ICD9_DGNS_CD_5='V5869', ICD9_DGNS_CD_6='42731', ICD9_DGNS_CD_7=None, ICD9_DGNS_CD_8=None, ICD9_DGNS_CD_9=None, ICD9_DGNS_CD_10=None, ICD9_PRCDR_CD_1=None, ICD9_PRCDR_CD_2=None, ICD9_PRCDR_CD_3=None, ICD9_PRCDR_CD_4=None, ICD9_PRCDR_CD_5=None, ICD9_PRCDR_CD_6=None, NCH_BENE_PTB_DDCTBL_AMT='0', NCH_BENE_PTB_COINSRNC_AMT='0', ADMTNG_ICD9_DGNS_CD=None, HCPCS_CD_1='85610', HCPCS_CD_2='80048', HCPCS_CD_3='80061', HCPCS_CD_4='82306', HCPCS_CD_5='96372', HCPCS_CD_6='87088', HCPCS_CD_7='85025', HCPCS_CD_8='80076', HCPCS_CD_9='84075', HCPCS_CD_10=None, HCPCS_CD_11=None, HCPCS_CD_12=None, HCPCS_CD_13=None, HCPCS_

In [7]:
df_selected = df_data.select("CLM_ID","ICD9_DGNS_CD_1","ICD9_DGNS_CD_2","ICD9_DGNS_CD_3","ICD9_DGNS_CD_4","Fraud")

from pyspark.sql.functions import col, when
k = col("ICD9_DGNS_CD_1")
df_selected = df_selected.withColumn("ICD9_DGNS_CD_1", when(k.isNull(), "__NA__").otherwise(k))

from pyspark.sql.functions import col, when
k = col("ICD9_DGNS_CD_2")
df_selected = df_selected.withColumn("ICD9_DGNS_CD_2", when(k.isNull(), "__NA__").otherwise(k))

from pyspark.sql.functions import col, when
k = col("ICD9_DGNS_CD_3")
df_selected = df_selected.withColumn("ICD9_DGNS_CD_3", when(k.isNull(), "__NA__").otherwise(k))

from pyspark.sql.functions import col, when
k = col("ICD9_DGNS_CD_4")
df_selected = df_selected.withColumn("ICD9_DGNS_CD_4", when(k.isNull(), "__NA__").otherwise(k))

df_selected.cache()
df_selected.persist()
df_selected.take(5)

[Row(CLM_ID='5.42272E+14', ICD9_DGNS_CD_1='V5832', ICD9_DGNS_CD_2='V5861', ICD9_DGNS_CD_3='2724', ICD9_DGNS_CD_4='3182', Fraud='N'),
 Row(CLM_ID='5.42282E+14', ICD9_DGNS_CD_1='9594', ICD9_DGNS_CD_2='E9174', ICD9_DGNS_CD_3='4019', ICD9_DGNS_CD_4='__NA__', Fraud='N'),
 Row(CLM_ID='5.42132E+14', ICD9_DGNS_CD_1='OTHER', ICD9_DGNS_CD_2='__NA__', ICD9_DGNS_CD_3='__NA__', ICD9_DGNS_CD_4='__NA__', Fraud='Y'),
 Row(CLM_ID='5.42322E+14', ICD9_DGNS_CD_1='OTHER', ICD9_DGNS_CD_2='__NA__', ICD9_DGNS_CD_3='__NA__', ICD9_DGNS_CD_4='__NA__', Fraud='Y'),
 Row(CLM_ID='5.42632E+14', ICD9_DGNS_CD_1='OTHER', ICD9_DGNS_CD_2='__NA__', ICD9_DGNS_CD_3='__NA__', ICD9_DGNS_CD_4='__NA__', Fraud='Y')]

Explore the loaded data by using the following Apache® Spark DataFrame methods:
-  print schema
-  print top ten records
-  count all records

As you can see, the data contains five fields. PRODUCT_LINE field is the one we would like to predict (label).

In [8]:
df_selected.printSchema()

root
 |-- CLM_ID: string (nullable = true)
 |-- ICD9_DGNS_CD_1: string (nullable = true)
 |-- ICD9_DGNS_CD_2: string (nullable = true)
 |-- ICD9_DGNS_CD_3: string (nullable = true)
 |-- ICD9_DGNS_CD_4: string (nullable = true)
 |-- Fraud: string (nullable = true)



In [9]:
df_selected.count()

67834

As you can see, the data set contains 67834 records.

<a id="model"></a>
## 3. Create an Apache® Spark machine learning model

In this section you will learn how to prepare data, create an Apache® Spark machine learning pipeline, and train a model.

### 3.1: Prepare data

Split your data into the train, test, and predict data sets:

In [13]:
splitted_data = df_selected.randomSplit([80.0,18.0,2.0])
train_data = splitted_data[0]
test_data = splitted_data[1]
predict_data = splitted_data[2]

test_data.cache()
train_data.cache()
predict_data.cache()

print ("Number of training records: " , str(train_data.count()))
print ("Number of testing records : " , str(test_data.count()))
print ("Number of predict records : " + str(predict_data.count()))



Number of training records:  54460
Number of testing records :  12072
Number of predict records : 1302


The data has been split into three data sets: 

-  The train data set, which is the largest group, is used for training.
-  The test data set is used for model evaluation and to test the assumptions of the model.
-  The predict data set is used for predictions.

### 3.2: Create pipeline and train a model

In this section you will create an Apache® Spark machine learning pipeline and then train the model.

Import the Apache® Spark machine learning packages that you'll need in the subsequent steps:

In [14]:
from pyspark.ml.feature import OneHotEncoder, StringIndexer, IndexToString, VectorAssembler
from pyspark.ml.classification import RandomForestClassifier
from pyspark.ml.evaluation import BinaryClassificationEvaluator
from pyspark.ml.evaluation import MulticlassClassificationEvaluator
from pyspark.ml.evaluation import RegressionEvaluator
from pyspark.ml import Pipeline, Model

Convert all the string fields to numeric ones by using the StringIndexer transformer:

In [15]:
stringIndexer_icd9_1 = StringIndexer(inputCol="ICD9_DGNS_CD_1", outputCol="ICD9_DGNS_CD_1_IX").setHandleInvalid("skip")
stringIndexer_icd9_2 = StringIndexer(inputCol="ICD9_DGNS_CD_2", outputCol="ICD9_DGNS_CD_2_IX").setHandleInvalid("skip")
stringIndexer_icd9_3 = StringIndexer(inputCol="ICD9_DGNS_CD_3", outputCol="ICD9_DGNS_CD_3_IX").setHandleInvalid("skip")
stringIndexer_icd9_4 = StringIndexer(inputCol="ICD9_DGNS_CD_4", outputCol="ICD9_DGNS_CD_4_IX").setHandleInvalid("skip")
stringIndexer_fraud = StringIndexer(inputCol="Fraud", outputCol="label").setHandleInvalid("skip")

Create a feature vector by combining all features together:

In [16]:
vectorAssembler = VectorAssembler(inputCols=["ICD9_DGNS_CD_1_IX", "ICD9_DGNS_CD_2_IX", "ICD9_DGNS_CD_3_IX", "ICD9_DGNS_CD_4_IX"], outputCol="features")

Next, define the estimators you want to use for classification. Random Forest is used in the following example.

In [17]:
rf = RandomForestClassifier(labelCol="label", featuresCol="features", maxBins=8192, numTrees=10)

Now build the pipeline. A pipeline consists of transformers and an estimator.

In [18]:
pipeline_rf = Pipeline(stages=[stringIndexer_icd9_1, stringIndexer_icd9_2, stringIndexer_icd9_3, stringIndexer_icd9_4, stringIndexer_fraud, vectorAssembler, rf])

Train your Random Forest model by using your pipeline with the train_data set.

In [19]:
model_rf = pipeline_rf.fit(train_data)

You can check your **model accuracy** now. Evaluate the model with the test_data set.

In [20]:
results = model_rf.transform(test_data)

Model Evaluation

In [21]:
test_data.printSchema()
train_data.printSchema()

root
 |-- CLM_ID: string (nullable = true)
 |-- ICD9_DGNS_CD_1: string (nullable = true)
 |-- ICD9_DGNS_CD_2: string (nullable = true)
 |-- ICD9_DGNS_CD_3: string (nullable = true)
 |-- ICD9_DGNS_CD_4: string (nullable = true)
 |-- Fraud: string (nullable = true)

root
 |-- CLM_ID: string (nullable = true)
 |-- ICD9_DGNS_CD_1: string (nullable = true)
 |-- ICD9_DGNS_CD_2: string (nullable = true)
 |-- ICD9_DGNS_CD_3: string (nullable = true)
 |-- ICD9_DGNS_CD_4: string (nullable = true)
 |-- Fraud: string (nullable = true)



<a id="persistence"></a>
## 4. Persist the model

In [22]:
selected_results = results.persist()

In [23]:
selected_results.show(10)

+-----------+--------------+--------------+--------------+--------------+-----+-----------------+-----------------+-----------------+-----------------+-----+--------------------+-------------+-----------+----------+
|     CLM_ID|ICD9_DGNS_CD_1|ICD9_DGNS_CD_2|ICD9_DGNS_CD_3|ICD9_DGNS_CD_4|Fraud|ICD9_DGNS_CD_1_IX|ICD9_DGNS_CD_2_IX|ICD9_DGNS_CD_3_IX|ICD9_DGNS_CD_4_IX|label|            features|rawPrediction|probability|prediction|
+-----------+--------------+--------------+--------------+--------------+-----+-----------------+-----------------+-----------------+-----------------+-----+--------------------+-------------+-----------+----------+
|5.42012E+14|          1745|        __NA__|        __NA__|        __NA__|    N|            189.0|              0.0|              0.0|              0.0|  0.0|     (4,[0],[189.0])|   [10.0,0.0]|  [1.0,0.0]|       0.0|
|5.42012E+14|          1748|          V160|         V7281|          2449|    N|            216.0|            276.0|            110.0|   

In [25]:
print ("Fraud model precision = {:.2f}.".format(selected_results.filter(selected_results.Fraud == 'Y').filter(selected_results.prediction == 1.0 ).count()/float(selected_results.count()) ))

Fraud model precision = 0.01.


In [26]:
evaluator = BinaryClassificationEvaluator(rawPredictionCol="prediction",labelCol="label",metricName="areaUnderROC")
auc = evaluator.evaluate(selected_results)
print ('Area under curve = {:.2f}.'.format(auc))
print ("Area under curve = ", str(auc*100) + '%')

# Evaluate the model by computing the accuracy & error
mc_evaluator = MulticlassClassificationEvaluator(labelCol="label",predictionCol="prediction", metricName="accuracy")
accuracy = mc_evaluator.evaluate(selected_results)*100
error = 100 - accuracy
print("Accuracy =  " +  str(accuracy) + '%')
print("Test Error = " +  str(error) + '%')

# Evaluate the model by computing the RMSE on the selected data
rg_evaluator = RegressionEvaluator(metricName="rmse", labelCol="label", predictionCol="prediction")
rmse = rg_evaluator.evaluate(selected_results)*100
print("Root-mean-square error = " + str(rmse) + '%')


Area under curve = 1.00.
Area under curve =  99.72771827917953%
Accuracy =  99.45965417867436%
Test Error = 0.5403458213256442%
Root-mean-square error = 7.350821867829804%


In this section you will learn how to store your pipeline and model in Watson Machine Learning repository by using Python client libraries.

First, you must import client libraries.

In [27]:
from repository.mlrepositoryclient import MLRepositoryClient
from repository.mlrepositoryartifact import MLRepositoryArtifact

Put your authentication information from your instance of the Watson Machine Learning service on Bluemix in the next cell. You can find your information on the **Service Credentials** tab of your service instance in Bluemix.

In [70]:
service_path = 'https://us-south.ml.cloud.ibm.com'
username = 'e7ad3869-9346-46ba-a6ab-293b55b99028'
password = '232065f5-2a18-4c91-aa8e-2d5b0d38df7d'
instance_id = '118386ba-432d-4147-94ac-d85978f44550'


Authorize the repository client:

In [71]:
ml_repository_client = MLRepositoryClient(service_path)
ml_repository_client.authorize(username, password)

Create the model artifact (abstraction layer).

In [72]:
model_artifact = MLRepositoryArtifact(model_rf, training_data=train_data, name="Claim Fraud Prediction Model")

**Tip**: The MLRepositoryArtifact method expects a trained model object, training data, and a model name. (It is this model name that is displayed by the Watson Machine Learning service).

### 4.1: Save the pipeline and model

Save pipeline and model artifacts to your Watson Machine Learning instance:

In [73]:
saved_model = ml_repository_client.models.save(model_artifact)

Get the saved model metadata from Watson Machine Learning. Use the *meta.available_props()* method to get the list of available props.

In [74]:
saved_model.meta.available_props()

dict_keys(['authorEmail', 'runtime', 'label', 'lastUpdated', 'modelType', 'pipelineVersionHref', 'trainingDataRef', 'creationTime', 'evaluationMetrics', 'authorName', 'evaluationMethod', 'inputDataSchema', 'trainingDataSchema', 'modelVersionHref', 'version'])

In [76]:
print ("modelType: " + saved_model.meta.prop("modelType"))
print ("trainingDataSchema: " + str(saved_model.meta.prop("trainingDataSchema")))
print ("creationTime: " + str(saved_model.meta.prop("creationTime")))
print ("modelVersionHref: " + saved_model.meta.prop("modelVersionHref"))
print ("label: " + saved_model.meta.prop("label"))

modelType: sparkml-model-2.3
trainingDataSchema: {'type': 'struct', 'fields': [{'type': 'string', 'name': 'CLM_ID', 'metadata': {}, 'nullable': True}, {'type': 'string', 'name': 'ICD9_DGNS_CD_1', 'metadata': {}, 'nullable': True}, {'type': 'string', 'name': 'ICD9_DGNS_CD_2', 'metadata': {}, 'nullable': True}, {'type': 'string', 'name': 'ICD9_DGNS_CD_3', 'metadata': {}, 'nullable': True}, {'type': 'string', 'name': 'ICD9_DGNS_CD_4', 'metadata': {}, 'nullable': True}, {'type': 'string', 'name': 'Fraud', 'metadata': {}, 'nullable': True}]}
creationTime: 2018-11-12 02:34:03.351000+00:00
modelVersionHref: https://us-south.ml.cloud.ibm.com/v2/artifacts/models/6899f9c4-bd22-4600-b9c1-a0f457e5f9e9/versions/0127200e-04fa-4998-9ea2-ce411d34034e
label: Fraud


**Tip**: **modelVersionHref** is our model unique indentifier in the Watson Machine Learning repository.

### 4.2: Load the model

Load the saved model from the specified instance of Watson Machine Learning:

In [77]:
loadedModelArtifact = ml_repository_client.models.get(saved_model.uid)

You can print for example model name to make sure that model has been loaded correctly.

In [83]:
print (str(loadedModelArtifact.name))

Claim Fraud Prediction Model


As you can see the name is correct. You have learned how save and load the model from Watson Machine Learning repository.

<a id="visualization"></a>
## 5. Predict locally and visualize

In this section you will learn how to score test data using loaded model and visualize the prediction results with the Plotly package.

### 5.1: Make a local prediction using previously loaded model and test data

Score the predict_data data set:

In [84]:
predictions = loadedModelArtifact.model_instance().transform(predict_data)

Preview the results by calling the *show()* method on the predictions DataFrame.

In [85]:
predictions.show(5)

+-----------+--------------+--------------+--------------+--------------+-----+-----------------+-----------------+-----------------+-----------------+-----+--------------------+-------------+-----------+----------+
|     CLM_ID|ICD9_DGNS_CD_1|ICD9_DGNS_CD_2|ICD9_DGNS_CD_3|ICD9_DGNS_CD_4|Fraud|ICD9_DGNS_CD_1_IX|ICD9_DGNS_CD_2_IX|ICD9_DGNS_CD_3_IX|ICD9_DGNS_CD_4_IX|label|            features|rawPrediction|probability|prediction|
+-----------+--------------+--------------+--------------+--------------+-----+-----------------+-----------------+-----------------+-----------------+-----+--------------------+-------------+-----------+----------+
|5.42012E+14|          1144|          4019|         30002|         V1582|    N|           2097.0|              1.0|            445.0|             40.0|  0.0|[2097.0,1.0,445.0...|   [10.0,0.0]|  [1.0,0.0]|       0.0|
|5.42012E+14|          2778|          4010|         V1046|        __NA__|    N|           3077.0|            228.0|            119.0|   

Tabulate a count to see which claim line is fraud:

In [86]:
predictions.select("prediction").groupBy("prediction").count().show()

+----------+-----+
|prediction|count|
+----------+-----+
|       0.0| 1201|
|       1.0|   11|
+----------+-----+



### 5.2: Sample visualization of data with Plotly package

In this subsection you will explore prediction results with Plotly, which is an online analytics and data visualization tool.

**Example**: First, you need to install required packages. You can do it by running the following code. Run it only one time.<BR><BR>
!pip install plotly --user <BR>
!pip install cufflinks --user

Import Plotly and other required packages.

In [87]:
import sys
import pandas
import plotly.plotly as py
from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot
import plotly.tools as tls
import plotly.graph_objs as go
init_notebook_mode(connected=True)
sys.path.append("".join([os.environ["HOME"]])) 

Convert the Apache Spark DataFrame to a Pandas DataFrame.

In [88]:
predictions_pdf = predictions.toPandas()

Plot a pie chart that shows the percentage of predicted fraudulent claims for the predict test dataset.

In [89]:
cumulative_stats = predictions_pdf.groupby(['prediction']).count()

fraud_data = [go.Pie(
            labels=cumulative_stats.index,
            values=cumulative_stats['label'],
    )]

fraud_layout = go.Layout(
    title='Fraud distribution',
)

fig = go.Figure(data=fraud_data, layout=fraud_layout)
iplot(fig)

With this data set, you might want to do some analysis of the probability of fraud per claim code by using a bar chart.

In [90]:

fraud_claims = predictions_pdf.query('prediction == 1.0')
print(fraud_claims.ICD9_DGNS_CD_1 + fraud_claims.ICD9_DGNS_CD_2 + fraud_claims.ICD9_DGNS_CD_3 + fraud_claims.ICD9_DGNS_CD_4)


222     OTHER__NA____NA____NA__
310      2139__NA____NA____NA__
454     OTHER__NA____NA____NA__
472      5186__NA____NA____NA__
528     OTHER__NA____NA____NA__
569     OTHER__NA____NA____NA__
681     OTHER__NA____NA____NA__
701     20420__NA____NA____NA__
1081    OTHER__NA____NA____NA__
1125    OTHER__NA____NA____NA__
1211    OTHER__NA____NA____NA__
dtype: object


Based on bar plot you created, you might make the following conclusion: The following ICD9 code has maximum chances of being a fraudulent claim

<a id="scoring"></a>
## 6. Deploy and score in a Cloud

In this section you will learn how to create online scoring and to score a new data record by using the Watson Machine Learning REST API. 
For more information about REST APIs, see the [Watson Machine Learning API Documentation](http://watson-ml-api.mybluemix.net/).

To work with the Watson Machine Leraning REST API, generate an access token:

In [91]:
import urllib3, requests, json

headers = urllib3.util.make_headers(basic_auth='{}:{}'.format(username, password))
url = '{}/v3/identity/token'.format(service_path)
response = requests.get(url, headers=headers)
mltoken = 'Bearer ' + json.loads(response.text).get('token')
print (mltoken)


Bearer eyJhbGciOiJSUzUxMiIsInR5cCI6IkpXVCJ9.eyJ0ZW5hbnRJZCI6IjExODM4NmJhLTQzMmQtNDE0Ny05NGFjLWQ4NTk3OGY0NDU1MCIsImluc3RhbmNlSWQiOiIxMTgzODZiYS00MzJkLTQxNDctOTRhYy1kODU5NzhmNDQ1NTAiLCJwbGFuSWQiOiIzZjZhY2Y0My1lZGU4LTQxM2EtYWM2OS1mOGFmM2JiMGNiZmUiLCJyZWdpb24iOiJ1cy1zb3V0aCIsInVzZXJJZCI6ImU3YWQzODY5LTkzNDYtNDZiYS1hNmFiLTI5M2I1NWI5OTAyOCIsImlzcyI6Imh0dHBzOi8vdXMtc291dGgubWwuY2xvdWQuaWJtLmNvbS92My9pZGVudGl0eSIsImlhdCI6MTU0MTk5MDIyNCwiZXhwIjoxNTQyMDE5MDI0LCJjcmVhdGVkVGltZSI6MTU0MTk5MDIyNH0.jVIZVurd_6FMYR-jfU4HUPcAMCXcrJ05pyMAuNiDXvAt5AZVRWsiL7Zlfozuk67r9IYBpZUojymvnOSnGDtX2bG8spdGAC_aLF3yhZAqrxdYrNTnabZyraoFQU8ujtbLMh74GqAE_M5KvUlGGLEK7hd-6EeLDgRqdUUoQYrKvq3rpDuxaS5MQAp7slr8IPK7R19xxS4iwofj2VsEfZ-F92V2Rd-joI9slFjgd3jVOEenfg6-mk0IGQb9En-kSd-6lwSVnx_8G8v5ZXOiCv2_NTnslB-bm2t2HU1KUdI6O4cQMR4j0VEK2xZ6EBCVKpxEhZyQkMOaVaeen0RAS4LL6w


### 6.1: Create an online scoring endpoint

Now you can create an online scoring endpoint. Execute the following sample code that uses the **modelVersionHref** value to create the scoring endpoint to the Bluemix repository. Do only one deployment

response_online = requests.post(endpoint_online, json=payload_online, headers=header_online)
print response_online
print response_online.text

scoring_href = json.loads(response_online.text).get('entity').get('scoringHref')

In [107]:

endpoint_instance = service_path + "/v3/wml_instances/" + instance_id
header = {'Content-Type': 'application/json', 'Authorization': mltoken}

response_get_instance = requests.get(endpoint_instance, headers=header)
print (response_get_instance)
print (response_get_instance.text)

endpoint_published_models = json.loads(response_get_instance.text).get('entity').get('published_models').get('url')
print (endpoint_published_models)

header = {'Content-Type': 'application/json', 'Authorization': mltoken}
response_get = requests.get(endpoint_published_models, headers=header)
 
print (response_get)
print (response_get.text)

[endpoint_deployments] = [x.get('entity').get('deployments').get('url') 
for x in json.loads(response_get.text).get('resources') if x.get('metadata').get('guid') == saved_model.uid]

print(endpoint_deployments)

payload_online = {"name": "Claim Fraud Prediction", "description": "My Claim Fraud Deployment", "type": "online"}
response_online = requests.post(endpoint_deployments, json=payload_online, headers=header)

scoring_url = json.loads(response_online.text).get('entity').get('scoring_url')
print (scoring_url)


<Response [200]>
{
  "metadata": {
    "guid": "118386ba-432d-4147-94ac-d85978f44550",
    "url": "https://us-south.ml.cloud.ibm.com/v3/wml_instances/118386ba-432d-4147-94ac-d85978f44550",
    "created_at": "2018-11-12T02:05:02.784Z",
    "modified_at": "2018-11-12T02:49:28.500Z"
  },
  "entity": {
    "source": "Bluemix",
    "published_models": {
      "url": "https://us-south.ml.cloud.ibm.com/v3/wml_instances/118386ba-432d-4147-94ac-d85978f44550/published_models"
    },
    "usage": {
      "expiration_date": "2018-12-01T00:00:00.000Z",
      "computation_time": {
        "limit": 180000,
        "current": 0
      },
      "model_count": {
        "limit": 200,
        "current": 2
      },
      "prediction_count": {
        "limit": 5000,
        "current": 0
      },
      "gpu_count": {
        "limit": 8,
        "current": 0
      },
      "capacity_units": {
        "limit": 180000000,
        "current": 0
      },
      "deployment_count": {
        "limit": 5,
        "cur

In [104]:
!pip install -U jsonpickle

  Using cached https://files.pythonhosted.org/packages/ca/ce/97404d5aeb58e6155c216825c81b50f6eca8a5345c582317ae48391878f8/jsonpickle-1.0-py2.py3-none-any.whl


In [111]:
claim_scoring_endpoint='https://us-south.ml.cloud.ibm.com/v3/wml_instances/118386ba-432d-4147-94ac-d85978f44550/published_models/6899f9c4-bd22-4600-b9c1-a0f457e5f9e9/deployments/6c897e2d-eb11-4d06-8491-f9cbb81c949f/online'
data = {}
data['fields']=["CLM_ID","ICD9_DGNS_CD_1","ICD9_DGNS_CD_2","ICD9_DGNS_CD_3","ICD9_DGNS_CD_4"]
data['values'] = [["TEST_FRAUD_TRUE","81313","__NA__","__NA__","__NA__"],["TEST_FRAUD_FALSE","1745","__NA__","__NA__","__NA__"]]
import jsonpickle
payload_online = jsonpickle.encode(data)
thawed = jsonpickle.decode(payload_online)
print (thawed)


response_online = requests.post(claim_scoring_endpoint, json=thawed, headers=header)
print (response_online)
print (response_online.text)


{'fields': ['CLM_ID', 'ICD9_DGNS_CD_1', 'ICD9_DGNS_CD_2', 'ICD9_DGNS_CD_3', 'ICD9_DGNS_CD_4'], 'values': [['TEST_FRAUD_TRUE', '81313', '__NA__', '__NA__', '__NA__'], ['TEST_FRAUD_FALSE', '1745', '__NA__', '__NA__', '__NA__']]}
<Response [200]>
{
  "fields": ["CLM_ID", "ICD9_DGNS_CD_1", "ICD9_DGNS_CD_2", "ICD9_DGNS_CD_3", "ICD9_DGNS_CD_4", "ICD9_DGNS_CD_1_IX", "ICD9_DGNS_CD_2_IX", "ICD9_DGNS_CD_3_IX", "ICD9_DGNS_CD_4_IX", "features", "rawPrediction", "probability", "prediction"],
  "values": [["TEST_FRAUD_TRUE", "81313", "__NA__", "__NA__", "__NA__", 3999.0, 0.0, 0.0, 0.0, [4, [0], [3999.0]], [8.0, 2.0], [0.8, 0.2], 0.0], ["TEST_FRAUD_FALSE", "1745", "__NA__", "__NA__", "__NA__", 189.0, 0.0, 0.0, 0.0, [4, [0], [189.0]], [10.0, 0.0], [1.0, 0.0], 0.0]]
}


<a id="summary"></a>
## 7. Summary and next steps     

 You successfully completed this notebook! You learned how to use Apache Spark machine learning as well as Watson Machine Learning for model creation and deployment. Check out our _[Online Documentation](https://apsportal.ibm.com/docs/content/analyze-data/ml-overview.html)_ for more samples, tutorials, documentation, how-tos, and blog posts. 

Copyright © 2017 IBM. This notebook and its source code are released under the terms of the MIT License.